docs/secp256k1/add-to-jni.md

```diff
@@ -0,0 +1,439 @@
+<!-- END doctoc generated TOC please keep comment here to allow auto update -->
+
+## Adding a new function to NativeSecp256k1.java
+
+ ### Adding to `src/java/org_bitcoin_NativeSecp256k1.c`
+
+1. Add an `#include` import at the top (if applicable)
+
+   If your secp256k1 functions are not already included, you will need to `#include` the header file (should be in the `secp256k1/include` directory).
+
+2. Function signature
+
+   Your new function signature should begin with
+
+   ```c
+   SECP256K1_API jint JNICALL Java_org_bitcoin_NativeSecp256k1_
+   ```
+
+   followed by the secp256k1 function name where `_` are replaced with `_1` (it's a weird jni thing). Finally, you add a parameter list that begins with
+
+   ```c
+   (JNIEnv* env, jclass classObject, jobject byteBufferObject, jlong ctx_l
+   ```
+
+   and ends with any `jint`s in the case that any of the secp256k1 function inputs have variable length (such as public keys which can be either `33` or `65` bytes, or ECDSA signatures), and lastly any `jboolean`s in case there is some flag like `compressed` passed in.
+
+   As an example that includes everything, if you are making a call to `secp256k1_pubkey_tweak_add` which takes in public keys that could be `33` or `65` bytes and outputs a public key that will either be compressed or decompressed based on an input flag, then the function signature would be
+
+   ```c
+   SECP256K1_API jobjectArray JNICALL Java_org_bitcoin_NativeSecp256k1_secp256k1_1pubkey_1tweak_1add
+     (JNIEnv* env, jclass classObject, jobject byteBufferObject, jlong ctx_l, jint publen, jboolean compressed)
+   ```
+
+3. Reading `unsigned char*` inputs
+
+   It is now time to create pointers for each of the secp256k1 function inputs that where passed in via the `byteBufferObject`. We must first read in the `Secp256k1Context` with the line
+
+   ```c
+   secp256k1_context *ctx = (secp256k1_context*)(uintptr_t)ctx_l;
+   ```
+
+   and we can then initialize the first pointer to be the beginning of the `byteBufferObject` with the line
+
+   ```c
+   unsigned char* firstArg = (*env)->GetDirectBufferAddress(env, byteBufferObject);
+   ```
+
+   and subsequent arguments' pointers where the previous argument's length is known (say `32` bytes for example) can be instantiated using
+
+   ```c
+   unsigned char* prevArg = ...
+   unsigned char* nextArg = (unsigned char*) (prevArg + 32);
+   ```
+
+   and in the case that a previous argument has variable length, then a `jint` has been provided as an input and can be used instead, such as in the example
+
+   ```c
+   unsigned char* prevArg = ...
+   unsigned char* nextArg = (unsigned char*) (prevArg + publen);
+   ```
+
+   where `publen` is a `jint` passed to this C function.
+
+   As an example that includes everything, consider the function `secp256k1_ecdsa_verify` which takes as input a `32` byte message, a variable length signature and a public key (of length `33` or `65` bytes). Our function will begin with
+
+   ```c
+   {
```

Comment:
unnecessary `{`

Reply:
Actually I meant for it to be here as I'm trying to spell out the entire function and this seemed like the better place to include the brace than in the function signature